code/datums/components/tether.dm

```diff
@@ -1,39 +1,171 @@
+	tether_beam = beam.visuals
 	if (ispath(tether_name, /atom))
 		var/atom/tmp = tether_name
 		src.tether_name = initial(tmp.name)
 	else
 		src.tether_name = tether_name
-	RegisterSignals(parent, list(COMSIG_MOVABLE_PRE_MOVE), PROC_REF(checkTether))
 
-/datum/component/tether/proc/checkTether(mob/mover, newloc)
+/datum/component/tether/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(check_tether))
+	RegisterSignal(tether_target, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(check_tether))
+	RegisterSignal(tether_target, COMSIG_QDELETING, PROC_REF(on_delete))
+	RegisterSignal(tether_beam, COMSIG_CLICK, PROC_REF(beam_click))
+	// Also snap if the beam gets deleted, more of a backup check than anything
+	RegisterSignal(tether_beam, COMSIG_QDELETING, PROC_REF(on_delete))
+
+	if (!isnull(embed_target))
+		RegisterSignal(embed_target, COMSIG_ITEM_UNEMBEDDED, PROC_REF(on_embedded_removed))
+		RegisterSignal(embed_target, COMSIG_QDELETING, PROC_REF(on_delete))
+
+/datum/component/tether/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_MOVABLE_PRE_MOVE)
+	if (!QDELETED(tether_target))
+		UnregisterSignal(tether_target, list(COMSIG_MOVABLE_PRE_MOVE, COMSIG_QDELETING))
+	if (!QDELETED(tether_beam))
+		UnregisterSignal(tether_beam, list(COMSIG_CLICK, COMSIG_QDELETING))
+		qdel(tether_beam)
+	if (!QDELETED(embed_target))
+		UnregisterSignal(embed_target, list(COMSIG_ITEM_UNEMBEDDED, COMSIG_QDELETING))
+
+/datum/component/tether/proc/check_tether(atom/source, new_loc)
 	SIGNAL_HANDLER
 
-	if (get_dist(mover,newloc) > max_dist)
-		to_chat(mover, span_userdanger("The [tether_name] runs out of slack and prevents you from moving!"))
+	if (!isturf(new_loc))
+		to_chat(source, span_warning("[tether_name] prevents you from entering [new_loc]!"))
 		return COMPONENT_MOVABLE_BLOCK_PRE_MOVE
 
+	var/atom/movable/anchor = (source == tether_target ? parent : tether_target)
+	if (get_dist(anchor, new_loc) > cur_dist)
+		if (!istype(anchor) || anchor.anchored || !anchor.Move(get_step_towards(anchor, new_loc)))
+			to_chat(source, span_warning("[tether_name] runs out of slack and prevents you from moving!"))
+			return COMPONENT_MOVABLE_BLOCK_PRE_MOVE
+
 	var/atom/blocker
-	out:
-		for(var/turf/T in get_line(tether_target,newloc))
-			if (T.density)
-				blocker = T
-				break out
-			for(var/a in T)
-				var/atom/A = a
-				if(A.density && A != mover && A != tether_target)
-					blocker = A
-					break out
+	for (var/turf/line_turf in get_line(anchor, new_loc))
+		if (line_turf.density && line_turf != anchor.loc && line_turf != source.loc)
+			blocker = line_turf
+			break
+		for (var/atom/in_turf in line_turf)
+			if (line_turf == anchor.loc || line_turf == source.loc)
+				if ((in_turf.flags_1 & ON_BORDER_1) && (in_turf.dir & get_dir(source, anchor)))
+					blocker = in_turf
+					break
+				continue
```

Comment:
if this if check changes the flow of the loop totally why not do it on the outside and swap between two for loops (I assume this is reasonably hot code)